tools/emscripten_wasmfs_fuzzer/workload.h

```diff
@@ -0,0 +1,166 @@
+// Copyright 2021 The Emscripten Authors.  All rights reserved.
+// Emscripten is available under two separate licenses, the MIT license and the
+// University of Illinois/NCSA Open Source License.  Both these licenses can be
+// found in the LICENSE file.
+
+// This file defines fuzzer workloads which primarily fuzz multi-threaded
+// operations.
+
+#pragma once
+
+#include <assert.h>
+#include <cstdlib>
+#include <iostream>
+#include <optional>
+#include <stdio.h>
+#include <string>
+#include <thread>
+#include <unistd.h>
+#include <unordered_map>
+
+namespace wasmfs {
+
+// This abstract class defines a fuzzer workload. It sets up a series of file
+// system operations and then validates that the expected outcome is observed.
+class Workload {
+protected:
+  Random& rand;
+
+public:
+  Workload(Random& rand) : rand(rand) {}
+  virtual ~Workload() = default;
+
+  // This function should only generate the Random data that seeds a workload.
+  // This could be a list of files to rename or random file content to write.
+  virtual void generate() = 0;
+
+  // This function defines the actual file system property that should be
+  // validated. Any file system syscalls should be called by this function.
+  virtual void execute() = 0;
+};
+
+// This workload class attempts to fuzz for tearing in reads and writes.
+// This should validate that writes are atomic (i.e. that a read interleaved
+// between two reads should not read a portion of the first read and a portion
+// of the second read). Writer threads should write a uniform string of the same
+// character. Reader threads should validate that file content is not intermixed
+// (it should check that all characters in the file content are the same).
+class ReadWrite : public Workload {
+  int fd;
+  std::vector<std::string> work;
+  std::unordered_map<std::thread::id, std::string> threads;
+  std::atomic<bool> go{false};
+
+  bool isSame(std::vector<char>& target) {
+    std::set<char> seen;
+
+    for (auto it = target.begin(); it != target.end() - 1; it++) {
+      seen.insert(*it);
+      if (seen.size() > 1) {
+#ifdef FUZZER_DEBUG
+        for (int i = 0; i < target.size(); i++) {
+          if (target[i] == 0) {
+            printf("\\0");
+          } else {
+            printf("%c", target[i]);
+          }
+        }
+#endif
+        return false;
+      }
+    }
+    return true;
+  }
+
+  void writer() {
+    while (!go) {
+    }
+
+    for (int i = 0; i < rand.upTo(10); i++) {
+#ifdef FUZZER_DEBUG
+      std::cout << std::this_thread::get_id() << " is running.\n";
+#endif
+      auto current = rand.upTo(work.size());
```

Comment:
If we are non-deterministically executing the generated `work`, then we might as well just generate the workload non-deterministically on the fly rather than in a separate generate step. That might simplify the code as well.